src/orion/algo/axoptimizer.py

```diff
@@ -0,0 +1,344 @@
+"""
+:mod:`orion.algo.axoptimizer` -- TODO
+========================================
+
+TODO: Write long description
+"""
+import contextlib
+import copy
+from typing import List, Optional
+
+try:
+    from ax.service.ax_client import AxClient
+    from ax.service.utils.instantiation import ObjectiveProperties
+
+    has_Ax = True
+except ImportError:
+    AxClient = None
+    ObjectiveProperties = None
+    has_Ax = False
+
+from orion.algo.base import BaseAlgorithm
+from orion.core.utils import format_trials
+from orion.core.utils.flatten import flatten
+
+
+class AxOptimizer(BaseAlgorithm):
+    """TODO: Class docstring
+
+    Parameters
+    ----------
+    space: `orion.algo.space.Space`
+        Optimisation space with priors for each dimension.
+    seed: None, int or sequence of int, optional
+        random seed for reproducibility. Works only for Sobol quasi-random
+        generator and for BoTorch-powered models. For the latter models, the
+        trials generated from the same optimization setup with the same seed,
+        will be mostly similar, but the exact parameter values may still vary
+        and trials latter in the optimizations will diverge more and more.  This
+        is because a degree of randomness is essential for high performance of
+        the Bayesian optimization models and is not controlled by the seed.
+
+        .. note:: In multi-threaded environments, the random seed is
+           thread-safe, but does not actually guarantee reproducibility. Whether
+           the outcomes will be exactly the same for two same operations that
+           use the random seed, depends on whether the threads modify the random
+           state in the same order across the two operations.  Default: ``None``
+    n_initial_trials: int, optional
+        Specific number of initialization trials.
+        Initialization trials are generated quasi-randomly using Sobol.
+    extra_objectives: sequence of str, optional
+        List of metrics' name which are also objectives to minimize.
+        threshold: The bound in the objective's threshold constraint.
+
+        .. note:: Orion expects the `extra_objectives` results to be stored in
+           `orion.core.worker.Trial.statistics`
+    constraints: sequence of str, optional
+        Dict of list of string representation of metrics constraints of form
+        ["metric_name >= bound"], like ["m1 <= 3"]
+
+        .. note:: Orion expects the `constraints` results to be stored in
+           `orion.core.worker.Trial.constraints`
+        .. note:: See https://ax.dev/docs/core.html#optimization-config for more
+           details about how Ax expects its outcome constraints
+    """
+
+    requires_type = "numerical"
+    requires_dist = None
+    requires_shape = "flattened"
+
+    def __init__(
+        self,
+        space,
```

Comment:
Might as well type-hint it with `Space`, right? :)
```suggestion
        space: Space,
```